packages/aws-cdk/lib/api/aws-auth/sdk-provider.ts

```diff
@@ -0,0 +1,369 @@
+    /**
+     * Default region
+     */
+    public readonly defaultRegion: string,
+    httpOptions: SdkHttpOptions = {}) {
+      this.httpOptions = defaultHttpOptions(httpOptions);
+    }
+
+  /**
+   * Return an SDK which can do operations in the given environment
+   *
+   * The `region` and `accountId` parameters are interpreted as in `resolveEnvironment()` (which is to
+   * say, `undefined` doesn't do what you expect).
+   */
+  public async forEnvironment(accountId: string | undefined, region: string | undefined, mode: Mode): Promise<ISDK> {
+    const env = await this.resolveEnvironment(accountId, region);
+    const creds = await this.obtainCredentials(env.account, mode);
+    return new SDK(creds, env.region, this.httpOptions);
+  }
+
+  /**
+   * Return an SDK which uses assumed role credentials
+   *
+   * The base credentials used to retrieve the assumed role credentials will be the
+   * current credentials (no plugin lookup will be done!).
+   *
+   * If `region` is undefined, the default value will be used.
+   */
+  public async withAssumedRole(roleArn: string, externalId: string | undefined, region: string | undefined) {
+    debug(`Assuming role '${roleArn}'`);
+    region = region ?? this.defaultRegion;
+
+    const creds = new AWS.ChainableTemporaryCredentials({
+      params: {
+        RoleArn: roleArn,
+        ...externalId ? { ExternalId: externalId } : {},
+        RoleSessionName: `aws-cdk-${os.userInfo().username}`,
+      },
+      stsConfig: {
+        region,
+        ...this.httpOptions,
+      },
+      masterCredentials: await this.defaultCredentials(),
+    });
+
+    return new SDK(creds, region, this.httpOptions);
+  }
+
+  /**
+   * Resolve the environment for a stack
+   *
+   * `undefined` actually means `undefined`, and is NOT changed to default values! Only the magic values UNKNOWN_REGION
+   * and UNKNOWN_ACCOUNT will be replaced with looked-up values!
+   */
+  public async resolveEnvironment(accountId: string | undefined, region: string | undefined) {
+    region = region !== cxapi.UNKNOWN_REGION ? region : this.defaultRegion;
+    accountId = accountId !== cxapi.UNKNOWN_ACCOUNT ? accountId : (await this.defaultAccount())?.accountId;
+
+    if (!region) {
+      throw new Error(`AWS region must be configured either when you configure your CDK stack or through the environment`);
+    }
+
+    if (!accountId) {
+      throw new Error(`Unable to resolve AWS account to use. It must be either configured when you define your CDK or through the environment`);
+    }
+
+    const environment: cxapi.Environment = {
+      region, account: accountId, name: cxapi.EnvironmentUtils.format(accountId, region)
+    };
+
+    return environment;
+  }
+
+  /**
+   * Use the default credentials to lookup our account number using STS.
+   *
+   * Uses a cache to avoid STS calls if we don't need 'em.
+   */
+  public defaultAccount(): Promise<Account | undefined> {
```

Comment:
same question: seems like this is a duplicate of a function we already have

Reply:
Same answer: https://github.com/aws/aws-cdk/pull/6565#discussion_r387870491

There could be some shared logic between CLI and `cdk-assets` tool, but we don't really have a good (3rd) place to put it, so for now it's duplicated between them. I *guess* we could make `@aws-cdk/sdk` or something, or `@aws-cdk/tools` where we put everything, but I distinctly recall you saying to not bother with that right now, so we're not.

It's also not ENTIRELY the same (though it could be), the SDK CLI is doing much more its best to look and behave like the AWS CLI, whereas `cdk-assets` run by itself behaves unapologetically only like the AWS SDK for JS does.